clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -555,6 +555,9 @@ private void close(KafkaChannel channel, boolean processOutstanding) {
 
         channel.disconnect();
 
+        //avoid inconsistent connection states, see KAFKA-5417
+        connected.remove(channel.id());
```

Comment:
I was debating whether `doClose` is the more appropriate spot for this since that is where we actually close the channel and add the id to the `disconnected` collection. We have this "closing" state below in which we await pending receives before closing the channel. I am not sure if it is better to allow a connection in that state to be reported as connected or not. It probably doesn't matter too much for the specific bug reported since we wouldn't have any pending receives, but I guess we should still make sure this the state transitions are consistent. Thoughts?